uwh/comms.py

```diff
@@ -0,0 +1,100 @@
+from . import messages_pb2
+from .gamemanager import GameState, TimeoutState
+
+def to_proto_enum(gamemanager_enum):
+    return { messages_pb2.GameState_GameOver        : GameState.game_over,
+             messages_pb2.GameState_FirstHalf       : GameState.first_half,
+             messages_pb2.GameState_HalfTime        : GameState.half_time,
+             messages_pb2.GameState_SecondHalf      : GameState.second_half,
+             messages_pb2.TimeoutState_RefTimeout   : TimeoutState.ref,
+             messages_pb2.TimeoutState_WhiteTimeout : TimeoutState.ref, # bold-faced lie
+             messages_pb2.TimeoutState_BlackTimeout : TimeoutState.ref # bold-faced lie
+           }[gamemanager_enum]
+
+
+def from_proto_enum(proto_enum):
+    return { GameState.game_over   : messages_pb2.GameState_GameOver,
+             GameState.first_half  : messages_pb2.GameState_FirstHalf,
+             GameState.half_time   : messages_pb2.GameState_HalfTime,
+             GameState.second_half : messages_pb2.GameState_SecondHalf,
+             TimeoutState.ref      : messages_pb2.TimeoutState_RefTimeout,
+           }[proto_enum]
+
+
+class UWHProtoHandler(object):
+    def __init__(self, mgr):
+        self._mgr = mgr
+
+    def send_message(self, recipient, msg_kind, msg):
+        self.send_raw(recipient, self.pack_message(msg_kind, msg))
+
+    def send_raw(self, recipient, data):
+        """ To be implemented by the deriving class """
+        raise NotImplementedError("Not Yet Implemented")
+
+    def recv_message(self, sender, kind, msg):
+        if kind == messages_pb2.MessageType_Ping:
+            self.handle_Ping(sender, msg)
+        elif kind == messages_pb2.MessageType_GameKeyFrame:
+            self.handle_GameKeyFrame(sender, msg)
+
+    def recv_raw(self, sender, data):
+        (kind, msg) = self.unpack_message(data)
+        self.recv_message(sender, kind, msg)
+
+    def message_for_msg_kind(self, msg_kind):
+        return { messages_pb2.MessageType_Ping : messages_pb2.Ping,
+                 messages_pb2.MessageType_Pong : messages_pb2.Pong,
+                 messages_pb2.MessageType_GameKeyFrame : messages_pb2.GameKeyFrame
+               }[msg_kind]()
+
+    def pack_message(self, msg_kind, msg):
+        length = msg.ByteSize()
+        assert length < 255 # XBee can't handle huge messages
```

Comment:
You might want to change these asserts to raising exceptions. And then catch them from the top-level loop and log that the unexpected happened.

Reply:
Good idea. That'll make problems non-fatal.